src/server/server.rs

```diff
@@ -307,67 +310,64 @@ impl<T: RaftStoreRouter> Server<T> {
         }
     }
 
-    fn connect_peer(&mut self,
-                    event_loop: &mut EventLoop<Self>,
-                    sock_addr: SocketAddr,
-                    peer: &str)
-                    -> Result<Token> {
+    fn connect_store(&mut self,
+                     event_loop: &mut EventLoop<Self>,
+                     store_id: u64,
+                     sock_addr: SocketAddr)
+                     -> Result<Token> {
+        // We may already create the connection before.
+        if let Some(token) = self.store_tokens.get(&store_id).cloned() {
+            warn!("token already exists for store {}, reuse", store_id);
+            return Ok(token);
+        }
+
         let sock = try!(TcpStream::connect(&sock_addr));
-        let token = try!(self.add_new_conn(event_loop, sock, Some(peer.to_owned())));
-        self.peers.insert(peer.to_owned(), token);
+        let token = try!(self.add_new_conn(event_loop, sock, Some(store_id)));
+        self.store_tokens.insert(store_id, token);
         Ok(token)
     }
 
-    fn resolve_peer(&mut self, peer: String, data: ConnData) {
-        // If the address is host:port, resolving the host's IP may
-        // block the event loop, so we should use an asynchronous way
-        // to resolve the host address and then send again.
-        let peer_addr = peer.clone();
+    fn resolve_store(&mut self, store_id: u64, data: ConnData) {
         let ch = self.sendch.clone();
         let cb = box move |r| {
             if let Err(e) = r {
-                error!("resolve peer {} err {:?}", peer, e);
+                error!("resolve store {} err {:?}", store_id, e);
                 return;
             }
 
-            let sock = r.unwrap();
-            if let Err(e) = ch.send(Msg::SendPeerSock {
-                sock_addr: sock,
-                peer: peer,
+            let addr = r.unwrap();
+            if let Err(e) = ch.send(Msg::SendStoreSock {
+                store_id: store_id,
+                sock_addr: addr,
                 data: data,
             }) {
-                error!("send peer sock msg err {:?}", e);
+                error!("send store sock msg err {:?}", e);
             }
         };
-        if let Err(e) = self.resolver.resolve(peer_addr, cb) {
+        if let Err(e) = self.resolver.resolve(store_id, cb) {
             error!("try to resolve err {:?}", e);
         }
     }
 
-    fn send_peer(&mut self, event_loop: &mut EventLoop<Self>, addr: String, data: ConnData) {
-        // check the corresponding token for peer address.
-        if let Some(token) = self.peers.get(&addr).cloned() {
+    fn send_store(&mut self, event_loop: &mut EventLoop<Self>, store_id: u64, data: ConnData) {
```

Comment:
It is safe here. 
If we don't find an existing token (which means the connection is ok, and we can send data with it) , if not we will use pd to get the address first.